File: implementations/java/provekit-claim-envelope/src/test/java/com/provekit/claimenvelope/CicpConformanceTest.java
// SPDX-License-Identifier: Apache-2.0

package com.provekit.claimenvelope;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

import org.junit.jupiter.api.Test;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.provekit.ir.Jcs;
import com.provekit.ir.Jcs.Value;

class CicpConformanceTest {
    private static final Path CICP_DIR = findCicpDir();
    private static final ObjectMapper JSON = new ObjectMapper();

    @Test
    void passing_vectors_derive_expected_blake3_512_cids() throws IOException {
        JsonNode catalog = JSON.readTree(CICP_DIR.resolve("vectors.json").toFile());

        for (JsonNode vector : catalog.path("vectors")) {
            if (!vector.path("shouldPass").asBoolean()) {
                continue;
            }

            JsonNode body = readBody(vector);
            assertEquals(List.of(), validateClosedInputs(body), vector.path("name").asText());
            assertEquals(
                vector.path("expectedCid").asText(),
                Jcs.blake3Cid(toJcsValue(body)),
                vector.path("name").asText());
        }
    }

    @Test
    void invalid_vectors_fail_closed_on_missing_input_cids() throws IOException {
        JsonNode catalog = JSON.readTree(CICP_DIR.resolve("vectors.json").toFile());

        for (JsonNode vector : catalog.path("vectors")) {
            if (vector.path("shouldPass").asBoolean()) {
                continue;
            }

            List<String> errors = validateClosedInputs(readBody(vector));
            assertFalse(errors.isEmpty(), vector.path("name").asText());
            assertTrue(
                errors.stream().anyMatch(error -> error.contains(vector.path("errorContains").asText())),
                vector.path("name").asText() + " errors: " + errors);
        }
    }

    private static JsonNode readBody(JsonNode vector) throws IOException {
        return JSON.readTree(CICP_DIR.resolve(vector.path("body").asText()).toFile());
    }

    private static Value toJcsValue(JsonNode node) {
        if (node.isObject()) {
            LinkedHashMap<String, Value> entries = new LinkedHashMap<>();
            node.fields().forEachRemaining(field -> entries.put(field.getKey(), toJcsValue(field.getValue())));
            return Value.object(entries);
        }
        if (node.isArray()) {
            List<Value> items = new ArrayList<>();
            node.forEach(item -> items.add(toJcsValue(item)));
            return Value.array(items);
        }
        if (node.isTextual()) {
            return Value.string(node.asText());
        }
        if (node.isIntegralNumber()) {
            return Value.integer(node.asLong());
        }
        if (node.isBoolean()) {
            return Value.bool(node.asBoolean());
        }
        if (node.isNull()) {
            return Value.NULL;
        }
        throw new IllegalArgumentException("Unsupported JSON node for JCS: " + node);
    }

    private static List<String> validateClosedInputs(JsonNode body) {
        Set<String> inputCids = textSet(body.path("inputCids"));
        Set<String> requiredCids = requiredInputCids(body);
        List<String> errors = new ArrayList<>();

        if (!body.path("inputCids").isArray()) {
            errors.add("inputCids missing required array");
            return errors;
        }
        for (String cid : requiredCids) {
            if (!inputCids.contains(cid)) {
                errors.add("inputCids missing required CID " + cid);
            }
        }
        return errors;
    }

    private static Set<String> requiredInputCids(JsonNode body) {
        Set<String> cids = new TreeSet<>();
        switch (body.path("kind").asText()) {
            case "CIBlastRadius" -> {
                addText(cids, body, "protocolCatalogCid");
                addText(cids, body, "jobDefinitionCid");
                addText(cids, body, "commandCid");
                addText(cids, body, "runnerIdentityCid");
                addText(cids, body, "sourceClosureCid");
                addText(cids, body, "policyCid");
                addTexts(cids, body, "toolchainCids");
                addTexts(cids, body, "lockfileCids");
                addTexts(cids, body, "generatedInputCids");
                addTexts(cids, body, "fixtureCids");
                addTexts(cids, body, "relevantSpecCids");
            }
            case "CIJobResultBodyClaim" -> {
                addText(cids, body, "blastRadiusCid");
                addText(cids, body, "outputCid");
                addText(cids, body, "logCid");
                addText(cids, body, "runnerIdentityCid");
                addText(cids, body, "policyCid");
            }
            case "CIReuseBodyClaim" -> {
                addText(cids, body, "currentBlastRadiusCid");
                addText(cids, body, "previousBlastRadiusCid");
                addText(cids, body, "previousResultWitnessCid");
                addText(cids, body, "policyCid");
                addTexts(cids, body, "bridgeWitnessCids");
            }
            case "CIImpactBodyClaim" -> {
                addText(cids, body, "baseStateCid");
                addText(cids, body, "candidateStateCid");
                addText(cids, body, "policyCid");
                addTexts(cids, body, "protocolEvolutionWitnessCids");
                addTexts(cids, body, "changedBlastRadiusCids");
                addTexts(cids, body, "unchangedBlastRadiusCids");
                addTexts(cids, body, "reusableWitnessCids");
                addTexts(cids, body, "refusalCids");
            }
            default -> throw new IllegalArgumentException("Unknown CICP body kind: " + body.path("kind").asText());
        }
        return cids;
    }

    private static Set<String> textSet(JsonNode node) {
        Set<String> values = new TreeSet<>();
        if (node.isArray()) {
            node.forEach(item -> {
                if (item.isTextual()) {
                    values.add(item.asText());
                }
            });
        }
        return values;
    }

    private static void addText(Set<String> cids, JsonNode body, String field) {
        JsonNode value = body.path(field);
        if (value.isTextual()) {
            cids.add(value.asText());
        }
    }

    private static void addTexts(Set<String> cids, JsonNode body, String field) {
        if (body.path(field).isArray()) {
            body.path(field).forEach(value -> {
                if (value.isTextual()) {
                    cids.add(value.asText());
                }
            });
        }
    }

    private static Path findCicpDir() {
        Path dir = Path.of("").toAbsolutePath();
        while (dir != null) {
            Path candidate = dir.resolve(Path.of("protocol", "conformance", "cicp"));
            if (Files.exists(candidate.resolve("vectors.json"))) {
                return candidate;
            }
            dir = dir.getParent();
        }
        throw new IllegalStateException("Could not locate protocol/conformance/cicp/vectors.json");
    }
}

File: implementations/go/provekit-ir-symbolic/canonicalizer/cicp_conformance_test.go
package canonicalizer

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

type cicpManifest struct {
	Vectors []cicpVector `json:"vectors"`
}

type cicpVector struct {
	Name          string `json:"name"`
	Body          string `json:"body"`
	ExpectedCID   string `json:"expectedCid"`
	ShouldPass    bool   `json:"shouldPass"`
	ErrorContains string `json:"errorContains"`
}

func TestCICPGoldenVectors(t *testing.T) {
	dir := cicpConformanceDir()
	manifest := readCICPManifest(t, filepath.Join(dir, "vectors.json"))

	for _, vector := range manifest.Vectors {
		vector := vector
		t.Run(vector.Name, func(t *testing.T) {
			body := readCICPJSON(t, filepath.Join(dir, vector.Body))

			if !vector.ShouldPass {
				err := validateCICPInputClosure(body)
				if err == nil {
					t.Fatalf("validateCICPInputClosure succeeded; want fail-closed error containing %q", vector.ErrorContains)
				}
				if !strings.Contains(err.Error(), vector.ErrorContains) {
					t.Fatalf("validateCICPInputClosure error = %q, want substring %q", err, vector.ErrorContains)
				}
				return
			}

			if err := validateCICPInputClosure(body); err != nil {
				t.Fatalf("validateCICPInputClosure: %v", err)
			}

			canonical, err := EncodeJCS(body)
			if err != nil {
				t.Fatalf("EncodeJCS(%s): %v", vector.Body, err)
			}
			got := ComputeCID(canonical)
			if got != vector.ExpectedCID {
				t.Fatalf("CID mismatch for %s:\n  got:  %s\n  want: %s", vector.Body, got, vector.ExpectedCID)
			}
		})
	}
}

func cicpConformanceDir() string {
	return filepath.Clean(filepath.Join("..", "..", "..", "..", "protocol", "conformance", "cicp"))
}

func readCICPManifest(t *testing.T, path string) cicpManifest {
	t.Helper()

	var manifest cicpManifest
	readJSONFile(t, path, &manifest)
	if len(manifest.Vectors) == 0 {
		t.Fatalf("%s: no vectors", path)
	}
	return manifest
}

func readCICPJSON(t *testing.T, path string) map[string]interface{} {
	t.Helper()

	var body map[string]interface{}
	readJSONFile(t, path, &body)
	return body
}

func readJSONFile(t *testing.T, path string, into interface{}) {
	t.Helper()

	file, err := os.Open(path)
	if err != nil {
		t.Fatalf("open %s: %v", path, err)
	}
	defer file.Close()

	dec := json.NewDecoder(file)
	dec.UseNumber()
	if err := dec.Decode(into); err != nil {
		t.Fatalf("decode %s: %v", path, err)
	}
}

func validateCICPInputClosure(body map[string]interface{}) error {
	inputs, err := stringSetField(body, "inputCids")
	if err != nil {
		return err
	}

	for _, cid := range requiredCICPInputCIDs(body) {
		if _, ok := inputs[cid]; !ok {
			return fmt.Errorf("inputCids missing required CID %s", cid)
		}
	}
	return nil
}

func requiredCICPInputCIDs(body map[string]interface{}) []string {
	switch stringField(body, "kind") {
	case "CIBlastRadius":
		return appendCIDFields(body,
			[]string{
				"protocolCatalogCid",
				"jobDefinitionCid",
				"commandCid",
				"runnerIdentityCid",
				"sourceClosureCid",
				"policyCid",
			},
			[]string{
				"toolchainCids",
				"lockfileCids",
				"generatedInputCids",
				"fixtureCids",
				"relevantSpecCids",
			},
		)
	case "CIJobResultBodyClaim":
		return appendCIDFields(body,
			[]string{
				"blastRadiusCid",
				"outputCid",
				"logCid",
				"runnerIdentityCid",
				"policyCid",
			},
			nil,
		)
	case "CIReuseBodyClaim":
		return appendCIDFields(body,
			[]string{
				"currentBlastRadiusCid",
				"previousBlastRadiusCid",
				"previousResultWitnessCid",
				"policyCid",
			},
			[]string{"bridgeWitnessCids"},
		)
	case "CIImpactBodyClaim":
		return appendCIDFields(body,
			[]string{
				"baseStateCid",
				"candidateStateCid",
				"policyCid",
			},
			[]string{
				"protocolEvolutionWitnessCids",
				"changedBlastRadiusCids",
				"unchangedBlastRadiusCids",
				"reusableWitnessCids",
				"refusalCids",
			},
		)
	default:
		return nil
	}
}

func appendCIDFields(body map[string]interface{}, scalarFields, listFields []string) []string {
	var cids []string
	for _, field := range scalarFields {
		if cid := stringField(body, field); cid != "" {
			cids = append(cids, cid)
		}
	}
	for _, field := range listFields {
		cids = append(cids, stringSliceField(body, field)...)
	}
	return cids
}

func stringField(body map[string]interface{}, field string) string {
	value, _ := body[field].(string)
	return value
}

func stringSliceField(body map[string]interface{}, field string) []string {
	values, ok := body[field].([]interface{})
	if !ok {
		return nil
	}
	out := make([]string, 0, len(values))
	for _, value := range values {
		if s, ok := value.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

func stringSetField(body map[string]interface{}, field string) (map[string]struct{}, error) {
	values, ok := body[field].([]interface{})
	if !ok {
		return nil, fmt.Errorf("%s must be an array", field)
	}

	out := make(map[string]struct{}, len(values))
	for i, value := range values {
		s, ok := value.(string)
		if !ok {
			return nil, fmt.Errorf("%s[%d] must be a string", field, i)
		}
		out[s] = struct{}{}
	}
	return out, nil
}
